backend/src/services/auth.service.ts

```
@@ -123,4 +133,40 @@ export const AuthService = {
       return errorResponse(ErrorCodes.INTERNAL_ERROR, String(error));
     }
   },
+
+  /**
+   * POST /auth/refresh
+   * Securely rotates tokens.
+   */
+  async refreshToken(token: string): Promise<ServiceResult<any>> {
+    const tokenRes = await BaseService.query('SELECT * FROM refresh_token WHERE token = $1', [token]);
+    const storedToken = tokenRes.rows[0];
+
+    if (!storedToken || storedToken.is_revoked) {
+      return errorResponse(ErrorCodes.UNAUTHORIZED, 'Invalid or revoked refresh token')
+    }
+
+    try {
+      const payload = jwt.verify(token, process.env.JWT_REFRESH_SECRET as string) as RefreshTokenPayload;
+
+      await BaseService.query('UPDATE refresh_token SET is_revoked = true WHERE token_id = $1', [storedToken.token_id]);
+
+      const newAccessToken = signAccessToken(payload.uid);
+      const newRefreshToken = signRefreshToken(payload.uid);
+      const newExpiresAt = getExpirationDate(newRefreshToken);
+
+      await BaseService.query(
+        `INSERT INTO refresh_token (user_id, token, expires_at) VALUES ($1, $2, $3)`,
+        [payload.uid, newRefreshToken, newExpiresAt]
+      );
+
+      return successResponse(SuccessCodes.OK, {
+        accessToken: newAccessToken,
+        refreshToken: newRefreshToken,
+      });
+    } catch (err) {
+      await BaseService.query('UPDATE refresh_token SET is_revoked = true WHERE token_id = $1', [storedToken.token_id]);
+      return errorResponse(ErrorCodes.UNAUTHORIZED, 'Invalid or revoked refresh token')
```

let's throw a different error here, because the token may have been valid but something else went wrong for example in the creation of the new token